lib/gen_stage/dispatchers/demand_dispatcher.ex

```diff
@@ -5,13 +5,30 @@ defmodule GenStage.DemandDispatcher do
   This is the default dispatcher used by `GenStage`. In order
   to avoid greedy consumers, it is recommended that all consumers
   have exactly the same maximum demand.
+
+  ## Options
+
+  The demand dispatcher accepts the following options
+  on initialization:
+
+    * `:shuffle_demands_on_first_dispatch` - when true, shuffle the initial demands list
+      which is constructed on subscription before first dispatch. It prevents overloading
+      the first consumer on first dispatch.  Defaults to `false`.
+
+  ### Examples
+
+  To start a producer with demands shuffled on first dispatch:
+
+      {:producer, state, dispatcher: {GenStage.DemandDispatcher, shuffle_demands_on_first_dispatch: true}}
   """
 
   @behaviour GenStage.Dispatcher
 
   @doc false
-  def init(_opts) do
-    {:ok, {[], 0, nil}}
+  def init(opts) do
+    shuffle_demand = Keyword.get(opts, :shuffle_demands_on_first_dispatch, false)
+
+    {:ok, {[], 0, nil, %{shuffle_demands_on_first_dispatch: shuffle_demand, demands_seq: :natural}}}
```

Comment:
So you mean the default value of the flag is `true` and we should always shuffle it?

Originally I thought we need an option to allow the library user, say Broadway, to set.  If we use this flag to only indicate whether shuffled or not, I think it can be a boolean.  My previous idea is to remember more info.

Reply:
No, what I mean is that all states that we have can be handled by a boolean. :)

If `:shuffle_demands_on_first_dispatch` is false, then the value in the tuple starts as false.

If `:shuffle_demands_on_first_dispatch` is true, then the value in the tuple starts as true and is switched to false after the first dispatch call.